lib/cn/spill.c

```diff
@@ -200,35 +200,58 @@ get_direct_read_buf(uint len, bool aligned_voff, u32 *bufsz, void **buf)
     return 0;
 }
 
-struct route_info {
-    struct route_node  *rnode;
-    struct key_obj      ekobj;
-    bool                last_node;
-    char                ekey[HSE_KVS_KEY_LEN_MAX];
+struct spill_ctx {
+    struct bin_heap *bh;
+
+    uint bldr_idx;
+    bool more;
+    struct merge_item curr;
+
+    /* Ptomb */
+    struct key_obj pt_kobj;
+    bool pt_set;
+
+    /* [HSE_REVISIT]: Maintain a list of seqnos for pt_kobj, not just one.
+     */
+    u64 pt_seq;
 };
 
-static void
-cn_spill_route_get(
-    struct cn_tree *tree,
-    struct key_obj *kobj,
-    struct route_info *ri)
+merr_t
+cn_spill_init(struct cn_compaction_work *w, void **ctx)
 {
-    unsigned klen;
+    struct spill_ctx *s;
+    merr_t err;
 
-    assert(sizeof(ri->ekey) == HSE_KVS_KEY_LEN_MAX);
+    s = malloc(sizeof(*s));
+    if (!s)
+        return merr(ENOMEM);
 
-    /* Borrow callers output buffer to convert input key obj to a char
-     * buffer as required by the route API.
-     */
-    key_obj_copy(ri->ekey, sizeof(ri->ekey), &klen, kobj);
+    err = merge_init(&s->bh, w->cw_inputv, w->cw_kvset_cnt, &w->cw_stats);
+    if (err)
+        return err;
+
+    memset(w->cw_outv, 0, w->cw_outc * sizeof(*w->cw_outv));
+
+    s->bldr_idx = 0;
+
+    memset(&s->pt_kobj, 0, sizeof(s->pt_kobj));
```

Comment:
if you memset this, might as well memset `s` or use `calloc` and then don't worry about initializing members to `0`.